diff --git a/usr/local/emhttp/plugins/dynamix.docker.manager/include/DockerClient.php b/usr/local/emhttp/plugins/dynamix.docker.manager/include/DockerClient.php
index 08215fd..2b69ffc 100644
--- a/usr/local/emhttp/plugins/dynamix.docker.manager/include/DockerClient.php
+++ b/usr/local/emhttp/plugins/dynamix.docker.manager/include/DockerClient.php
@@ -10,6 +10,7 @@
  * The above copyright notice and this permission notice shall be included in
  * all copies or substantial portions of the Software.
  */
+$composemanDockerClientPatch = true;
 ?>
 <?
 libxml_use_internal_errors(true); # Suppress any warnings from xml errors.
@@ -295,16 +296,29 @@ class DockerTemplates {
 			$tmp['paused'] = $ct['Paused'];
 			$tmp['autostart'] = in_array($name, $autoStart);
 			$tmp['cpuset'] = $ct['CPUset'];
-			if (!is_file($tmp['icon']) || $reload) $tmp['icon'] = $this->getIcon($image,$name);
+			// read docker label for WebUI & Icon
+			if ($ct['Url'] && !$tmp['url']) $tmp['url'] = $ct['Url'];
+			if ($ct['Icon']) $tmp['icon'] = $ct['Icon'];
+			if ($ct['Shell']) $tmp['shell'] = $ct['Shell'];
+
+			if (!is_file($tmp['icon']) || $reload) $tmp['icon'] = $this->getIcon($image,$name,$ct['Icon']);
+
 			if ($ct['Running']) {
 				$port = &$ct['Ports'][0];
-				$ip = ($ct['NetworkMode']=='host'||$port['NAT'] ? $host : $port['IP']);
-				$tmp['url'] = $ip ? (strpos($tmp['url'],$ip)!==false ? $tmp['url'] : $this->getControlURL($ct, $ip)) : $tmp['url'];
+				$webui = $tmp['url'] ?: $this->getTemplateValue($ct['Image'], 'WebUI');
+				if (strlen($webui) > 0 && !preg_match("%\[(IP|PORT:(\d+))\]%", $webui)) {
+					// non-templated webui, user specified
+					$tmp['url'] = $webui;
+				} else {
+					$ip = ($ct['NetworkMode']=='host'||$port['NAT'] ? $host : $port['IP']);
+					$tmp['url'] = $ip ? (strpos($tmp['url'],$ip)!==false ? $tmp['url'] : $this->getControlURL($ct, $ip)) : $tmp['url'];
+				}
 				$tmp['shell'] = $tmp['shell'] ?? $this->getTemplateValue($image, 'Shell');
 			}
 			$tmp['registry'] = $tmp['registry'] ?? $this->getTemplateValue($image, 'Registry');
 			$tmp['Support'] = $tmp['Support'] ?? $this->getTemplateValue($image, 'Support');
 			$tmp['Project'] = $tmp['Project'] ?? $this->getTemplateValue($image, 'Project');
+			if (strpos($ct['Manager'],"composeman")!==false) $tmp['updated'] = 'undef';
 			if (!$tmp['updated'] || $reload) {
 				if ($reload) $DockerUpdate->reloadUpdateStatus($image);
 				$tmp['updated'] = var_export($DockerUpdate->getUpdateStatus($image),true);
@@ -318,14 +332,16 @@ class DockerTemplates {
 		return $info;
 	}
 
-	public function getIcon($Repository,$contName) {
+	public function getIcon($Repository,$contName,$iconUrl='') {
 		global $docroot, $dockerManPaths;
-		$imgUrl = $this->getTemplateValue($Repository, 'Icon','all',$contName);
-		if (!$imgUrl) return '';
+		$imgUrl = $iconUrl ?: $this->getTemplateValue($Repository, 'Icon','all',$contName);
+		if (!$imgUrl || trim($imgUrl) == "/plugins/dynamix.docker.manager/images/question.png") return '';
+
+		$imgUrlHash = sha1($imgUrl);
+		$iconFile = sprintf('%s-%s.png', 'icon', $imgUrlHash);
 
-		$imageName = $contName ?: $name;
-		$iconRAM = sprintf('%s/%s-%s.png', $dockerManPaths['images-ram'], $contName, 'icon');
-		$icon    = sprintf('%s/%s-%s.png', $dockerManPaths['images'], $contName, 'icon');
+		$iconRAM = sprintf('%s/%s-%s', $dockerManPaths['images-ram'], $contName, $iconFile);
+		$icon    = sprintf('%s/%s', $dockerManPaths['images'], $iconFile);
 		
 		if (!is_dir(dirname($iconRAM))) mkdir(dirname($iconRAM), 0755, true);
 		if (!is_dir(dirname($icon))) mkdir(dirname($icon), 0755, true);
@@ -335,10 +351,47 @@ class DockerTemplates {
 			@copy($icon, $iconRAM);
 		}
 		if (!is_file($icon) && is_file($iconRAM)) {
-			@copy($iconRAM,$icon);
+			@copy($iconRAM, $icon);
 		}
+		if ( !is_file($iconRAM) ) {
+			exec("/usr/bin/logger ".escapeshellarg("$contName: Could not download icon $imgUrl"));
+		}
+		else {
+			$this->purgeUnusedIconFiles($contName, $iconFile);
+		}
+
 		return (is_file($iconRAM)) ? str_replace($docroot, '', $iconRAM) : '';
 	}
+
+	public function purgeUnusedIconFiles($contName, $keepIcon='') {
+		global $docroot, $dockerManPaths;
+
+		$icon_glob = sprintf('%s/%s-*.png', $dockerManPaths['images-ram'], $contName);
+		$ramFiles = glob($icon_glob);
+		foreach ($ramFiles as $filename) {
+			if ( ($keepIcon === '') || !(strpos($filename, $keepIcon) !== false) ) {
+				@unlink($filename);
+			}
+		}
+
+		// $icon_glob = sprintf('%s/%s*.png', $dockerManPaths['images'], $contName);
+		// foreach (glob($icon_glob) as $filename) {
+		// 	if ( ($keepIcon === '') || !(strpos($filename, $keepIcon) !== false) ) {
+		// 		@unlink($filename);
+		// 	}
+		// }	
+
+		$icon_glob = sprintf('%s/%s*.png', $dockerManPaths['images'], $contName);
+		foreach ($ramFiles as $ramFile) {
+			if ( strpos($ramFile, '-icon-') !== false ) {
+				$suffix = end(explode('-', $ramFile));
+				if ( !glob($dockerManPaths['images-ram'].'/*icon-'.$suffix) ) {
+					$filename = sprintf('%s/icon-%s', $dockerManPaths['images'], $suffix);
+					@unlink($filename);
+				}
+			}
+		}
+	}
 }
 
 ####################################
@@ -880,6 +933,10 @@ class DockerClient {
 			[$net, $id]       = explode(':',$c['NetworkMode']);
 			$c['CPUset']      = $info['HostConfig']['CpusetCpus'];
 			$c['BaseImage']   = $ct['Labels']['BASEIMAGE'] ?? false;
+			$c['Icon']        = $info['Config']['Labels']['net.unraid.docker.icon'] ?? false;
+			$c['Url']         = $info['Config']['Labels']['net.unraid.docker.webui'] ?? false;
+			$c['Shell']       = $info['Config']['Labels']['net.unraid.docker.shell'] ?? false;
+			$c['Manager']  	  = $info['Config']['Labels']['net.unraid.docker.managed'] ?? false;
 			$c['Ports']       = [];
 			if ($id) $c['NetworkMode'] = $net.str_replace('/',':',DockerUtil::ctMap($id)?:'/???');
 			if ($driver[$c['NetworkMode']]=='bridge') {
